README.md
# Multi Link Text

A Flutter package multi_link_text allows you to create text with clickable links and additional styling options.

<img width="379" alt="screen" src="https://github.com/therasuldev/multi-link-text/assets/74558294/f1982df8-e6c3-46c6-baaf-3c4f0e9a2b86">

## Features

**Clickable Links**
- `MultiLinkText` allows you to specify certain words as clickable links, and you can define actions to be performed when those links are clicked.

**Rich Text Support**
- It supports rich text formatting, enabling you to customize the appearance of different words in the text.

**Paragraph Handling**
- You can include paragraph breaks in the text, and the widget will handle them accordingly.


## Installation

To use this package, add **multi_link_text** as a dependency in your `pubspec.yaml` file.

```bash
dependencies:
  flutter:
    sdk: flutter
  multi_link_text: ^0.0.5  # Use the latest version
```
‼️ Then run the following command to fetch the package:

```bash
flutter clean
flutter pub get
```


## Usage


```dart
import 'package:flutter/material.dart';
import 'package:multi_link_text/multi_link_text.dart';

void main() {
  runApp(const MultiLinkTextExample());
}

class MultiLinkTextExample extends StatelessWidget {
  const MultiLinkTextExample({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Multi Link Text example'),
          backgroundColor: Colors.orange,
        ),
        body: const Center(
          child: MultiLinkText(
            data: 'It is Multi Link Text package. How to use it? Click ME !',
            params: {'ME': 'https://www.google.com'},
            extraParams: ['Multi', 'Link', 'Text', 'How'],
            paragraph: ['How'],
            extraParamsStyle: TextStyle(
              decoration: TextDecoration.underline,
              decorationStyle: TextDecorationStyle.wavy,
              color: Colors.green,
              fontSize: 25,
            ),
            paramsStyle: TextStyle(
              decoration: TextDecoration.underline,
              decorationStyle: TextDecorationStyle.wavy,
              color: Colors.orange,
              fontSize: 20,
              decorationColor: Color(0xFFFF6F00),
            ),
            style: TextStyle(fontSize: 25, color: Colors.black),
          ),
        ),
      ),
    );
  }
}


```

## Contributing

Contributions are welcome! Please feel free to submit issues and pull requests.


CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* TODO: Bug fix.

## 0.0.3

* TODO: Bug fix {image}.

## 0.0.4

* TODO: repository URL doesn't exist.

## 0.0.5

* TODO: code refactoring.

test/multi_link_text_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('multi link text', () {});
}


lib/multi_link_text.dart
library multi_link_text;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

class MultiLinkText extends StatefulWidget {
  const MultiLinkText({
    super.key,
    required this.data,
    this.extraParams,
    required this.params,
    this.style,
    this.paramsStyle = const TextStyle(color: Colors.blue),
    this.extraParamsStyle,
    this.paragraph,
    this.locale,
  });

  /// This should include the entire text.
  final String data;

  /// List of non-clickable words that are only visually differentiated using styles.
  final List<String>? extraParams;

  /// The map containing words as keys and corresponding URLs as values.
  final Map<String, String> params;

  /// The style of the [data]
  final TextStyle? style;

  /// The style of the words specified in [params].
  final TextStyle? paramsStyle;

  /// The style of the words specified in [extraParams].
  final TextStyle? extraParamsStyle;

  /// List of words to start a new paragraph.
  final List<String>? paragraph;

  /// The language of the text in this span and its span children.
  ///
  /// Setting the locale of this text span affects the way that assistive
  /// technologies, such as VoiceOver or TalkBack, pronounce the text.
  ///
  /// If this span contains other text span children, they also inherit the
  /// locale from this span unless explicitly set to different locales.
  final Locale? locale;

  @override
  State<MultiLinkText> createState() => _MultiLinkTextState();
}

class _MultiLinkTextState extends State<MultiLinkText> {
  // Function to extract words from the input text and return them in a list format.
  List<String> _params() {
    final textParams = widget.data.trim().split(RegExp(r'\s+'));
    return textParams;
  }

  // Function to determine the style of a word based on its type.
  TextStyle? _style({required String paramKey}) {
    // Check if the word is a clickable word (exists in widget.params).
    if (widget.params.keys
        .any((key) => RegExp(r'\b' + key + r'\b').hasMatch(paramKey))) {
      return widget.paramsStyle; // Return style for clickable words.
    }
    // Check if the word is a non-clickable word (exists in widget.extraParams).
    else if (widget.extraParams != null &&
        widget.extraParams!.contains(paramKey)) {
      return widget.extraParamsStyle; // Return style for non-clickable words.
    }
    // In this case, the word doesn't match any predefined styles, so it won't receive any specific style.
    return null;
  }

  // Function to create a tap gesture recognizer for clickable words in the "params" category.
  TapGestureRecognizer? _recognizer({required String word}) {
    return TapGestureRecognizer()
      ..onTap = () async {
        // Check if the word is a clickable word in the "params" category.
        if (widget.params.keys.contains(word)) {
          // Check if the corresponding URL is not empty.
          if (widget.params[word]!.isEmpty) return;

          // If the conditions are met, open the specified URL in a browser.
          await _goToBrowser(url: widget.params[word]!);
        }
      };
  }

  // Function to format the text, adding new lines for specified paragraph words.
  String _text({required String word}) {
    String text = '';

    // Check if the word is in the list of words that should start a new paragraph.
    if (widget.paragraph != null && widget.paragraph!.contains(word)) {
      // If the word should start a new paragraph, add a new line before the word.
      text += '\n$word ';
    } else {
      // If the word shouldn't start a new paragraph, add a space after the word.
      text += '$word ';
    }

    // Return the formatted text.
    return text;
  }

  Future<void> _goToBrowser({required String url}) async {
    try {
      await launchUrl(Uri.parse(url));
    } on Exception catch (exception) {
      throw exception.toString();
    }
  }

  @override
  void initState() {
    super.initState();
    _textParams = _params();
  }

  late List<String> _textParams;

  @override
  Widget build(BuildContext context) {
    return Text.rich(
      TextSpan(
        style: widget.style,
        children: List.generate(
          _textParams.length,
          (index) {
            final word = _textParams[index];
            return TextSpan(
              text: _text(word: word),
              style: _style(paramKey: word),
              recognizer: _recognizer(word: word),
              locale: widget.locale,
            );
          },
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:multi_link_text/multi_link_text.dart';

void main() {
  runApp(const MultiLinkTextExample());
}

class MultiLinkTextExample extends StatelessWidget {
  const MultiLinkTextExample({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Multi Link Text example'),
          backgroundColor: Colors.orange,
        ),
        body: const Center(
          child: MultiLinkText(
            data: 'It is Multi Link Text package. How to use it? Click ME !',
            params: {'ME': 'https://www.google.com'},
            extraParams: ['Multi', 'Link', 'Text', 'How'],
            paragraph: ['How'],
            extraParamsStyle: TextStyle(
              decoration: TextDecoration.underline,
              decorationStyle: TextDecorationStyle.wavy,
              color: Colors.green,
              fontSize: 25,
            ),
            paramsStyle: TextStyle(
              decoration: TextDecoration.underline,
              decorationStyle: TextDecorationStyle.wavy,
              color: Colors.orange,
              fontSize: 20,
              decorationColor: Color(0xFFFF6F00),
            ),
            style: TextStyle(fontSize: 25, color: Colors.black),
          ),
        ),
      ),
    );
  }
}
